core/validatorapi/grpc/validator.go

```diff
@@ -0,0 +1,512 @@
+	scheduler.WaitChainStart(server.Context(), s.eth2Cl, clockwork.NewRealClock())
+	return nil
+}
+
+// WaitForActivation 🕰 requires scheduler epoch events, 🎉 but can then just query validatorAPI.
+func (s *BeaconNodeValidatorServer) WaitForActivation(req *prysmpb.ValidatorActivationRequest, server prysmpb.BeaconNodeValidator_WaitForActivationServer) error {
+	done := make(chan struct{})
+	s.subscribeEpoch(func(epoch eth2p0.Epoch) {
+		vals, _ := s.eth2Cl.ValidatorsByPubKey(server.Context(), "head", toGroupPubkeys(req.PublicKeys))
+		for _, val := range vals {
+			if val.Status == eth2v1.ValidatorStateActiveOngoing {
+				close(done)
+			}
+		}
+	})
+
+	select {
+	case <-server.Context().Done():
+		return server.Context().Err()
+	case <-done:
+		return nil
+	}
+}
+
+// ValidatorIndex 🎉 can proxy to valiatgorAPI
+func (s *BeaconNodeValidatorServer) ValidatorIndex(ctx context.Context, req *prysmpb.ValidatorIndexRequest) (*prysmpb.ValidatorIndexResponse, error) {
+	resp, _ := s.valsProvider.ValidatorsByPubKey(ctx, "head", toGroupPubkeys([][]byte{req.PublicKey}))
+
+	return &prysmpb.ValidatorIndexResponse{Index: pb.ValidatorIndex(resp[0].Index)}, nil
+}
+
+// ValidatorStatus 🎉 can proxy to valiatgorAPI, 🙈 but will not return some data either not used in prysm or just used for logging.
+func (s *BeaconNodeValidatorServer) ValidatorStatus(ctx context.Context, req *prysmpb.ValidatorStatusRequest) (*prysmpb.ValidatorStatusResponse, error) {
+	resp, _ := s.valsProvider.ValidatorsByPubKey(ctx, "head", toGroupPubkeys([][]byte{req.PublicKey}))
+	return &prysmpb.ValidatorStatusResponse{
+		Status:                    prysmpb.ValidatorStatus(resp[0].Status),
+		Eth1DepositBlockNumber:    0, // TODO(corver): We do not have access to this, but it seems unused.
+		DepositInclusionSlot:      0, // TODO(corver): We do not have access to this, but it seems unused.
+		ActivationEpoch:           pb.Epoch(resp[0].Validator.ActivationEpoch),
+		PositionInActivationQueue: 0, // TODO(corver): We do not have access to this, it is only used for logging.
+	}, nil
+}
+
+// MultipleValidatorStatus 🎉 can proxy to valiatgorAPI, 🙈 but will not return some data either not used in prysm or just used for logging.
+func (s *BeaconNodeValidatorServer) MultipleValidatorStatus(ctx context.Context, req *prysmpb.MultipleValidatorStatusRequest) (*prysmpb.MultipleValidatorStatusResponse, error) {
+	// TODO(corver): This is same as ValidatorStatus above
+	return nil, nil
+}
+
+// GetBeaconBlock 🎉 can proxy to valiatgorAPI, 🔧 but requires access to config flag BuilderAPI.
+func (s *BeaconNodeValidatorServer) GetBeaconBlock(ctx context.Context, req *prysmpb.BlockRequest) (*prysmpb.GenericBeaconBlock, error) {
+	var randao eth2p0.BLSSignature
+	copy(randao[:], req.RandaoReveal)
+
+	if s.builderAPI {
+		block, _ := s.eth2Cl.BlindedBeaconBlockProposal(ctx, eth2p0.Slot(req.Slot), randao, req.Graffiti)
+		return convertBlindedBlock(block), nil
```

Comment:
this is interesting, why prysm would have generic beacon block object? why not deal with separate types?

Reply:
they have separate types, they just use a generic endpoint and wrapper